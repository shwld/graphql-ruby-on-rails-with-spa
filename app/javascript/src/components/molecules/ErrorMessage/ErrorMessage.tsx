import { ApolloError } from '@apollo/client'
import React from 'react'

export const ErrorMessage: React.FC<{
  error?: Pick<ApolloError, 'graphQLErrors'>
  columnName: string
}> = ({ error, columnName }) => {
  if (!error || !error.graphQLErrors || error.graphQLErrors.length === 0) {
    return <></>
  }
  const recordInvalidErrors = error.graphQLErrors.filter(
    (it) => it.extensions?.code === 'RECORD_INVALID'
  )
  /** HACK: Railsのエラーオブジェクトそのまま使っているが、ちょっと構造が複雑でが苦しい感じある */
  const errors = recordInvalidErrors
    .map(
      (it) =>
        (it.extensions?.record?.errors as {
          [columnName: string]: string[]
        }[]) || []
    )
    .flat()
  return (
    <>
      {errors
        .filter((it) => it[columnName])
        .map((it, i) => (
          <p key={i} className="text-red-500 text-xs italic">
            {(it[columnName] || []).map((e: string, j: number) => (
              <React.Fragment key={j}>
                {e}
                <br />
              </React.Fragment>
            ))}
          </p>
        ))}
    </>
  )
}
