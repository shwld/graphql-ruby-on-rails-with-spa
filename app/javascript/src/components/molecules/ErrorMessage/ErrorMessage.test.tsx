import React from 'react'
import { create } from 'react-test-renderer'

import { ErrorMessage } from './ErrorMessage'

describe('Molecules/ErrorMessage', () => {
  it('Snap Shot', () => {
    const component = create(
      <ErrorMessage error={{ graphQLErrors: [] }} columnName="name" />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
