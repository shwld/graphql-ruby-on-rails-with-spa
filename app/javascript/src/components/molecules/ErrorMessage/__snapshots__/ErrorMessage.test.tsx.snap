// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/ErrorMessage Snap Shot 1`] = `null`;
